src/python/pants/rules/core/list_roots.py

```diff
@@ -0,0 +1,59 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import Set
+
+from pants.engine.console import Console
+from pants.engine.fs import PathGlobs, Snapshot
+from pants.engine.goal import Goal, LineOriented
+from pants.engine.rules import console_rule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.source.source_root import AllExistingSourceRoots, SourceRoot, SourceRootConfig
+
+
+class Roots(LineOriented, Goal):
+  """List the repo's registered source roots."""
+  name = 'roots'
+
+
+@rule(AllExistingSourceRoots, [SourceRootConfig])
+def all_roots(source_root_config):
+
+  source_roots = source_root_config.get_source_roots()
+
+  all_paths: Set[str] = set()
+  for path in source_roots.traverse():
+    if path.startswith("^/"):
+      all_paths.add(f"{path[2:]}/")
+    else:
+      all_paths.add(f"**/{path}/")
+
+  snapshot = yield Get(Snapshot, PathGlobs(include=tuple(all_paths)))
+
+  all_source_roots: Set[SourceRoot] = set()
+  # Get(Snapshot, PathGlobs) picks up subdirectories of the source root
```

Comment:
Oh, I see, so if we glob on `**/src/*/` then it could pick up `src/python/src/foo` as a source root for foo? Is that the concern? Or does it match on every subdir? Because if the former then I think it would be better to rephrase as something like "The globs above can match on subdirs of source roots, e.g., src/py/foo/src/bar, so we use find_by_path to verify every candidate source root."

That is, it's important to explain precisely what the failure mode is here.